In [1]:
import numpy as np
import pandas as pd

In [12]:
books=pd.read_csv('books.csv')# Some line of books.csv might have an error so we have to skip these lines 

C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [13]:
books=pd.read_csv('books.csv',error_bad_lines=False,encoding='latin1')

C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [14]:
books

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [15]:
books=books[['ISBN','Book-Title','Book-Author','Year-Of-Publication','Publisher']]
books

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  
0                                Oxford University Press  
1                                  HarperFlamingo Canada  
2                                        HarperPerennial  
3                                   Farrar Straus Giroux  
4                             W. W. Norton &amp; Company  
...                                                  ...  
271355                   Random House Childrens Pub (Mm)  
271356                                      Dutton Books  
271357                                HarperSanFrancisco  
271358                           Oxford University Press  
271359  McGraw-Hill Humanities/Social Sciences/Languages  

[271360 rows x 5 columns]

In [24]:
# Renaming the columns
books=books.rename(columns={'Book-Title':'title','Book-Author':'author','Year-Of-Publication':'year','Publisher':'publisher'})

In [25]:
books

ISBN                                              title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                      author  year  \
0         Mark P. O. Morford  2002   
1       Richard Bruce Wright  2001   
2               Carlo D'Este  1991   
3           Gina Bari Kolata  1999   
4            E. J. W. Barber  1999   
...                      ...   ...   
271355        Paula Danziger  1988   
271356            Teri Sloat  1991   
271357      Christine Wicker  2004   
271358                 Plato  1996   
271359   Christopher  Biffle  2000   

                                               publisher  
0                                Oxford University Press  
1                                  HarperFlamingo Canada  
2                                        HarperPerennial  
3                                   Farrar Straus Giroux  
4                             W. W. Norton &amp; Company  
...                                                  ...  
271355                   Random House Childrens Pub (Mm)  
271356                                      Dutton Books  
271357                                HarperSanFrancisco  
271358                           Oxford University Press  
271359  McGraw-Hill Humanities/Social Sciences/Languages  

[271360 rows x 5 columns]

In [28]:
books.isnull().sum()#Very small amount of null values so will remove them after other dataframe are done

ISBN         0
title        0
author       1
year         0
publisher    2
dtype: int64

In [35]:
users=pd.read_csv('Users.csv',error_bad_lines=False,encoding='latin1')

C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [36]:
users

User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [37]:
users.rename(columns={'User-ID':'user_id','Location':'location','Age':'age'},inplace=True)

In [38]:
users

user_id                            location   age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [39]:
ratings=pd.read_csv('Ratings.csv',error_bad_lines=True,encoding='latin1')

C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [40]:
ratings

User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

In [44]:
ratings.rename(columns={"User-ID":'user_id',"Book-Rating":"rating"},inplace=True)

In [45]:
ratings

user_id         ISBN  rating
0         276725   034545104X       0
1         276726   0155061224       5
2         276727   0446520802       0
3         276729   052165615X       3
4         276729   0521795028       6
...          ...          ...     ...
1149775   276704   1563526298       9
1149776   276706   0679447156       0
1149777   276709   0515107662      10
1149778   276721   0590442449      10
1149779   276723  05162443314       8

[1149780 rows x 3 columns]

In [55]:
ratings.user_id.nunique()

105283

In [51]:
books.shape

(271360, 5)

In [53]:
users.shape

(278858, 3)

In [54]:
ratings.shape

(1149780, 3)

### First Problem
This is a huge dataset but the problem is that there are many books which have no ratings or very few user have rated them.

So will try to remove the books which have less than 50 users review (Hyperparameter: Can choose any number of users)


### Second Problem
Will remove the users who have given less than 200 book reviews (Hyperparameter)

In [65]:
ratings.ISBN[ratings['rating']==0].value_counts()

0971880107     1921
0316666343      588
0060928336      412
0385504209      396
044023722X      366
               ... 
0064461637        1
0078822246        1
013615302X        1
01400.5183X       1
0679752714        1
Name: ISBN, Length: 246724, dtype: int64

In [67]:
count=ratings.user_id.value_counts()
count

11676     13602
198711     7550
153662     6109
98391      5891
35859      5850
          ...  
116180        1
116166        1
116154        1
116137        1
276723        1
Name: user_id, Length: 105283, dtype: int64

In [76]:
y=count[count>200].index
y

Int64Index([ 11676, 198711, 153662,  98391,  35859, 212898, 278418,  76352,
            110973, 235105,
            ...
            260183,  73681,  44296, 155916,   9856, 274808,  28634,  59727,
            268622, 188951],
           dtype='int64', length=899)

In [79]:
#Now we only want these users to be in our dataframe
ratings[ratings['user_id'].isin(y)]

user_id        ISBN  rating
1456      277427  002542730X      10
1457      277427  0026217457       0
1458      277427  003008685X       8
1459      277427  0030615321       0
1460      277427  0060002050       0
...          ...         ...     ...
1147612   275970  3829021860       0
1147613   275970  4770019572       0
1147614   275970   896086097       0
1147615   275970  9626340762       8
1147616   275970  9626344990       0

[526356 rows x 3 columns]

In [80]:
ratings=ratings[ratings['user_id'].isin(y)]
ratings

user_id        ISBN  rating
1456      277427  002542730X      10
1457      277427  0026217457       0
1458      277427  003008685X       8
1459      277427  0030615321       0
1460      277427  0060002050       0
...          ...         ...     ...
1147612   275970  3829021860       0
1147613   275970  4770019572       0
1147614   275970   896086097       0
1147615   275970  9626340762       8
1147616   275970  9626344990       0

[526356 rows x 3 columns]

In [82]:
# Rating records reduce from from 1149780 to 526356 by eliminating non-useful users
#Only 899 users writes 526356 review

In [83]:
#Now we will join books and ratings Dataframe

In [95]:
ratings_with_books=ratings.merge(books,on='ISBN')

In [96]:
ratings_with_books

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
487666   275970  1892145022       0   
487667   275970  1931868123       0   
487668   275970  3411086211      10   
487669   275970  3829021860       0   
487670   275970  4770019572       0   

                                                    title  \
0       Politically Correct Bedtime Stories: Modern Ta...   
1       Politically Correct Bedtime Stories: Modern Ta...   
2       Politically Correct Bedtime Stories: Modern Ta...   
3       Politically Correct Bedtime Stories: Modern Ta...   
4       Politically Correct Bedtime Stories: Modern Ta...   
...                                                   ...   
487666                                   Here Is New York   
487667  There's a Porcupine in My Outhouse: Misadventu...   
487668                                         Die Biene.   
487669                                     The Penis Book   
487670                                            Musashi   

                                     author  year  \
0                         James Finn Garner  1994   
1                         James Finn Garner  1994   
2                         James Finn Garner  1994   
3                         James Finn Garner  1994   
4                         James Finn Garner  1994   
...                                     ...   ...   
487666                          E. B. White  1999   
487667                         Mike Tougias  2002   
487668  Sybil Gr�¤fin Sch�¶nfeldt  1993   
487669                         Joseph Cohen  1999   
487670                       Eiji Yoshikawa  1995   

                                   publisher  
0                  John Wiley &amp; Sons Inc  
1                  John Wiley &amp; Sons Inc  
2                  John Wiley &amp; Sons Inc  
3                  John Wiley &amp; Sons Inc  
4                  John Wiley &amp; Sons Inc  
...                                      ...  
487666                       Little Bookroom  
487667                    Capital Books (VA)  
487668  Bibliographisches Institut, Mannheim  
487669                              Konemann  
487670          Kodansha International (JPN)  

[487671 rows x 7 columns]

In [97]:
ratings_with_books.isnull().sum()

user_id      0
ISBN         0
rating       0
title        0
author       0
year         0
publisher    2
dtype: int64

In [99]:
books.shape

(271360, 5)

#### Points to be noted
1. ratings data frame records 526356, ratings_with_books records 487671    
2. books dataframe records 271360, ratings_with_books records 487671
3. So you can see records have decreased with respect to ratings data frame AS books DATAFRAME DON'T HAVE ALL BOOKS AS IN ratings DATAFRAME (THOSE BOOKS IN ratings DATAFRAME WHO WERE NOT IN books DATAFRAME WERE REMOVED) THAT IS WHY THE RECORDS DECREASED
4. So you can see records have increased with respect to books data frame

In [110]:
books.nunique()

ISBN         271360
title        242135
author       102023
year            202
publisher     16807
dtype: int64

In [113]:
ratings_with_books.nunique() # now a books ISBN number is repeated multiple time as multiple user have rated it.
#same for title and author etc....

user_id         899
ISBN         178018
rating           11
title        160269
author        67165
year            183
publisher     11008
dtype: int64

In [114]:
users

user_id                            location   age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

### We will take into account only those book who have been rated by atleast 50 user

In [160]:
# If you want to do this by earlier method then before merging ""ratings=ratings[ratings['user_id'].isin(y)]""
# and books dataframe apply conditon on ISBN also as z=ratings['ISBN'].value_counts(), z=z[z>=50], z.index
# ratings=ratings[ratings['ISBN'].isin(z)]
# ratings.merge(books,on="ISBN")
number_rating=ratings_with_books.groupby('title')['rating'].count().reset_index()  
number_rating

title  rating
0        A Light in the Storm: The Civil War Diary of ...       2
1                                   Always Have Popsicles       1
2                    Apple Magic (The Collector's series)       1
3        Beyond IBM: Leadership Marketing and Finance ...       1
4        Clifford Visita El Hospital (Clifford El Gran...       1
...                                                   ...     ...
160264  Ã?Â?ber die Pflicht zum Ungehorsam gegen den...       3
160265                                    Ã?Â?lpiraten.       1
160266                   Ã?Â?rger mit Produkt X. Roman.       1
160267                            Ã?Â?stlich der Berge.       1
160268                                Ã?Â?thique en toc       1

[160269 rows x 2 columns]

In [150]:
number_rating=number_rating.rename(columns={'rating':'number_of_ratings'})
number_rating

title  number_of_ratings
0        A Light in the Storm: The Civil War Diary of ...                  2
1                                   Always Have Popsicles                  1
2                    Apple Magic (The Collector's series)                  1
3        Beyond IBM: Leadership Marketing and Finance ...                  1
4        Clifford Visita El Hospital (Clifford El Gran...                  1
...                                                   ...                ...
160264  Ã?Â?ber die Pflicht zum Ungehorsam gegen den...                  3
160265                                    Ã?Â?lpiraten.                  1
160266                   Ã?Â?rger mit Produkt X. Roman.                  1
160267                            Ã?Â?stlich der Berge.                  1
160268                                Ã?Â?thique en toc                  1

[160269 rows x 2 columns]

### Now we have to merge ot into the ratings_with_books Dataframe

In [151]:
ratings_with_books.head(2)

user_id        ISBN  rating  \
0   277427  002542730X      10   
1     3363  002542730X       0   

                                               title             author  year  \
0  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   
1  Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner  1994   

                   publisher  
0  John Wiley &amp; Sons Inc  
1  John Wiley &amp; Sons Inc

In [152]:
final_df=ratings_with_books.merge(number_rating,on='title')

In [153]:
final_df

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
487666   275970  1892145022       0   
487667   275970  1931868123       0   
487668   275970  3411086211      10   
487669   275970  3829021860       0   
487670   275970  4770019572       0   

                                                    title  \
0       Politically Correct Bedtime Stories: Modern Ta...   
1       Politically Correct Bedtime Stories: Modern Ta...   
2       Politically Correct Bedtime Stories: Modern Ta...   
3       Politically Correct Bedtime Stories: Modern Ta...   
4       Politically Correct Bedtime Stories: Modern Ta...   
...                                                   ...   
487666                                   Here Is New York   
487667  There's a Porcupine in My Outhouse: Misadventu...   
487668                                         Die Biene.   
487669                                     The Penis Book   
487670                                            Musashi   

                                     author  year  \
0                         James Finn Garner  1994   
1                         James Finn Garner  1994   
2                         James Finn Garner  1994   
3                         James Finn Garner  1994   
4                         James Finn Garner  1994   
...                                     ...   ...   
487666                          E. B. White  1999   
487667                         Mike Tougias  2002   
487668  Sybil Gr�¤fin Sch�¶nfeldt  1993   
487669                         Joseph Cohen  1999   
487670                       Eiji Yoshikawa  1995   

                                   publisher  number_of_ratings  
0                  John Wiley &amp; Sons Inc                 82  
1                  John Wiley &amp; Sons Inc                 82  
2                  John Wiley &amp; Sons Inc                 82  
3                  John Wiley &amp; Sons Inc                 82  
4                  John Wiley &amp; Sons Inc                 82  
...                                      ...                ...  
487666                       Little Bookroom                  1  
487667                    Capital Books (VA)                  1  
487668  Bibliographisches Institut, Mannheim                  1  
487669                              Konemann                  1  
487670          Kodansha International (JPN)                  1  

[487671 rows x 8 columns]

In [154]:
ratings_with_books.shape,final_df.shape

((487671, 7), (487671, 8))

##### We will take into account only those book who have been rated by atleast 50 user

In [159]:
final_df=final_df[final_df['number_of_ratings']>=50]
final_df

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
236701   255489  0553579983       7   
236702   256407  0553579983       0   
236703   257204  0553579983       0   
236704   261829  0553579983       0   
236705   273979  0553579983       0   

                                                    title             author  \
0       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
...                                                   ...                ...   
236701                                   And Then You Die      Iris Johansen   
236702                                   And Then You Die      Iris Johansen   
236703                                   And Then You Die      Iris Johansen   
236704                                   And Then You Die      Iris Johansen   
236705                                   And Then You Die      Iris Johansen   

        year                  publisher  number_of_ratings  
0       1994  John Wiley &amp; Sons Inc                 82  
1       1994  John Wiley &amp; Sons Inc                 82  
2       1994  John Wiley &amp; Sons Inc                 82  
3       1994  John Wiley &amp; Sons Inc                 82  
4       1994  John Wiley &amp; Sons Inc                 82  
...      ...                        ...                ...  
236701  1998                     Bantam                 50  
236702  1998                     Bantam                 50  
236703  1998                     Bantam                 50  
236704  1998                     Bantam                 50  
236705  1998                     Bantam                 50  

[61853 rows x 8 columns]

In [172]:
#### If a same user have rated a book multiple times this can be bad for our model so we will clear the duplicates.
final_df=final_df.drop_duplicates(['title','user_id'])
final_df

user_id        ISBN  rating  \
0        277427  002542730X      10   
1          3363  002542730X       0   
2         11676  002542730X       6   
3         12538  002542730X      10   
4         13552  002542730X       0   
...         ...         ...     ...   
236701   255489  0553579983       7   
236702   256407  0553579983       0   
236703   257204  0553579983       0   
236704   261829  0553579983       0   
236705   273979  0553579983       0   

                                                    title             author  \
0       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
1       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
2       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
3       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
4       Politically Correct Bedtime Stories: Modern Ta...  James Finn Garner   
...                                                   ...                ...   
236701                                   And Then You Die      Iris Johansen   
236702                                   And Then You Die      Iris Johansen   
236703                                   And Then You Die      Iris Johansen   
236704                                   And Then You Die      Iris Johansen   
236705                                   And Then You Die      Iris Johansen   

        year                  publisher  number_of_ratings  
0       1994  John Wiley &amp; Sons Inc                 82  
1       1994  John Wiley &amp; Sons Inc                 82  
2       1994  John Wiley &amp; Sons Inc                 82  
3       1994  John Wiley &amp; Sons Inc                 82  
4       1994  John Wiley &amp; Sons Inc                 82  
...      ...                        ...                ...  
236701  1998                     Bantam                 50  
236702  1998                     Bantam                 50  
236703  1998                     Bantam                 50  
236704  1998                     Bantam                 50  
236705  1998                     Bantam                 50  

[59850 rows x 8 columns]

In [175]:
final_df.nunique()# In ratings_with_books user_id: 899 and in final_df it decrease to user_id : 888
# As that was those users which rated on books , WHO WERE RATED BY LESS THAN 50 USERS

user_id               888
ISBN                 2144
rating                 11
title                 742
author                579
year                   75
publisher             357
number_of_ratings     130
dtype: int64

In [180]:
book_pivot=final_df.pivot_table(values='rating',
    index='title',
    columns='user_id')
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
84 Charing Cross Road                                  NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     7.0   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   NaN  ...     NaN   
1st to Die: A Novel                                    NaN  ...     NaN   
2nd Chance                                             NaN  ...     NaN   
4 Blondes                                              0.0  ...     NaN   
84 Charing Cross Road                                  NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        NaN  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

##### fill NaN value 

In [182]:
book_pivot.fillna(0,inplace=True)
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [184]:
# As you can see this is a very sparse type Mostly filled with zeros.
# We don,t want to calculate distance from zeros as it will increse our computational load.

In [185]:
#So we will convert it into sparse_matrix  so that internally distance from zeros will not be calculated
from scipy.sparse import csr_matrix
book_sparse=csr_matrix(book_pivot)

In [187]:
list(book_sparse)

[<1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 28 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 47 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 34 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 13 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 25 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 26 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 15 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 25 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 5 stored 

In [183]:
#Will use KNN(classification and regression) brother Nearest Neighbour Algorithm that do Clustering.

In [189]:
from sklearn.neighbors import NearestNeighbors

In [192]:
model=NearestNeighbors(algorithm='brute')#n_neighbors=5 hyperparameter
#algorithm='auto'  'auto' will attempt to decide the most appropriate algorithm   hyperparameter
#algorithm='brute' We instruct the algo to calculate distance of all the points and then decide neighbors  

In [193]:
model.fit(book_sparse)

NearestNeighbors(algorithm='brute')

In [211]:
book_pivot.iloc[0,:].values.reshape(1,-1)

array([[ 9.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  8.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  9.,  0.,
         0.,  0.,  0.,  8.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  7.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  9.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        10.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
         0.,  9.,  0.,  0.,  0.,  0.,  0.,  0.,  0.

In [242]:
distances,suggestion=model.kneighbors(book_pivot.iloc[237,:].values.reshape(1, -1),n_neighbors=6)
#reshaping is must for our model to understand the data
#Neighbors=6 because 1 book will be the 237 book itself   AGAIN IT IS HYPERPARAMETER

In [243]:
distances#0----> book 237 distance from book 237

array([[ 0.        , 68.78953409, 69.5413546 , 72.64296249, 76.83098333,
        77.28518616]])

In [244]:
suggestion

array([[237, 240, 238, 241, 184, 536]], dtype=int64)

In [245]:
suggestion[0][0]

237

In [246]:
book_pivot.index[suggestion[0][0]]

'Harry Potter and the Chamber of Secrets (Book 2)'

In [275]:
# THE SUGGESTION FOR BOOK 237----> 'Harry Potter and the Chamber of Secrets (Book 2)' ARE
for i in range(len(suggestion)):
    print(book_pivot.index[suggestion[i]].values)

['Harry Potter and the Prisoner of Azkaban (Book 3)'
 'Harry Potter and the Goblet of Fire (Book 4)'
 "Harry Potter and the Sorcerer's Stone (Book 1)" 'Exclusive'
 'The Cradle Will Fall']


In [250]:
book_pivot.index[237]

'Harry Potter and the Chamber of Secrets (Book 2)'

In [258]:
book_pivot.loc['Harry Potter and the Chamber of Secrets (Book 2)',:]

user_id
254       9.0
2276      0.0
2766      0.0
2977      0.0
3363      0.0
         ... 
275970    9.0
277427    0.0
277478    0.0
277639    0.0
278418    0.0
Name: Harry Potter and the Chamber of Secrets (Book 2), Length: 888, dtype: float64

In [266]:
book_pivot.index=='Harry Potter and the Chamber of Secrets (Book 2)'

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [272]:
#np.where()  IS A FUNCTION THAT EXTRACT THE Values which is true
np.where(book_pivot.index=='Harry Potter and the Chamber of Secrets (Book 2)')[0][0]

237

In [307]:
def suggest(book_name):
    book_index=np.where(book_pivot.index==book_name)[0][0]
    distances,suggestion=model.kneighbors(book_pivot.iloc[book_index,:].values.reshape(1, -1),n_neighbors=6)
    for i in range(len(suggestion)):
        return 'The suggestion for the books: '+ str(book_pivot.index[suggestion[0][0]])+' are',book_pivot.index[suggestion[i]].values[1:6]

In [308]:
suggest('Harry Potter and the Chamber of Secrets (Book 2)')

('The suggestion for the books: Harry Potter and the Chamber of Secrets (Book 2) are',
 array(['Harry Potter and the Prisoner of Azkaban (Book 3)',
        'Harry Potter and the Goblet of Fire (Book 4)',
        "Harry Potter and the Sorcerer's Stone (Book 1)", 'Exclusive',
        'The Cradle Will Fall'], dtype=object))

In [309]:
suggest('4 Blondes')

('The suggestion for the books: 4 Blondes are',
 array(['No Safe Place', 'Pleading Guilty', 'Long After Midnight',
        'Exclusive', 'Lake Wobegon days'], dtype=object))

#### Files to save as pickle

##### 1.book_pivot
##### 2.model (Nearest Neighbors)

In [319]:
import pickle
pickle.dump(book_pivot,open('book_pivot.pkl','wb'))

In [320]:
pickle.dump(model,open('model.pkl','wb'))

In [315]:
image=pd.read_csv('books.csv',error_bad_lines=False,encoding='latin1')
image

C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
C:\Users\adila\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3441: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [428]:
image['Image-URL-M']

0         http://images.amazon.com/images/P/0195153448.0...
1         http://images.amazon.com/images/P/0002005018.0...
2         http://images.amazon.com/images/P/0060973129.0...
3         http://images.amazon.com/images/P/0374157065.0...
4         http://images.amazon.com/images/P/0393045218.0...
                                ...                        
271355    http://images.amazon.com/images/P/0440400988.0...
271356    http://images.amazon.com/images/P/0525447644.0...
271357    http://images.amazon.com/images/P/006008667X.0...
271358    http://images.amazon.com/images/P/0192126040.0...
271359    http://images.amazon.com/images/P/0767409752.0...
Name: Image-URL-M, Length: 271360, dtype: object

In [338]:
book_list=list(book_pivot.index.values)
book_list

['1984',
 '1st to Die: A Novel',
 '2nd Chance',
 '4 Blondes',
 '84 Charing Cross Road',
 'A Bend in the Road',
 'A Case of Need',
 'A Child Called \\It\\": One Child\'s Courage to Survive"',
 'A Civil Action',
 'A Cry In The Night',
 'A Darkness More Than Night',
 'A Day Late and a Dollar Short',
 'A Fine Balance',
 'A Great Deliverance',
 'A Heartbreaking Work of Staggering Genius',
 'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
 'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
 'A Man Named Dave: A Story of Triumph and Forgiveness',
 'A Man in Full',
 'A Map of the World',
 'A Painted House',
 'A Patchwork Planet',
 'A Prayer for Owen Meany',
 'A Thin Dark Line (Mysteries &amp; Horror)',
 "A Thousand Acres (Ballantine Reader's Circle)",
 'A Time to Kill',
 "A Virtuous Woman (Oprah's Book Club (Paperback))",
 'A Walk to Remember',
 'A Widow for One Year',
 'A Wrinkle In Time',
 'A Wrinkle in Time',
 'A Year in Provence',
 "ANGELA'S ASHES",
 'Abduction',
 'Abou

In [339]:
len(book_list)

742

In [341]:
book_list=pd.DataFrame(book_list,columns=['title'])
book_list

title
0                                                 1984
1                                  1st to Die: A Novel
2                                           2nd Chance
3                                            4 Blondes
4                                84 Charing Cross Road
..                                                 ...
737                                    Year of Wonders
738                                   You Belong To Me
739  Zen and the Art of Motorcycle Maintenance: An ...
740                                               Zoya
741                                \O\" Is for Outlaw"

[742 rows x 1 columns]

In [344]:
book_list['title'].values

array(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       '84 Charing Cross Road', 'A Bend in the Road', 'A Case of Need',
       'A Child Called \\It\\": One Child\'s Courage to Survive"',
       'A Civil Action', 'A Cry In The Night',
       'A Darkness More Than Night', 'A Day Late and a Dollar Short',
       'A Fine Balance', 'A Great Deliverance',
       'A Heartbreaking Work of Staggering Genius',
       'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
       'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
       'A Man Named Dave: A Story of Triumph and Forgiveness',
       'A Man in Full', 'A Map of the World', 'A Painted House',
       'A Patchwork Planet', 'A Prayer for Owen Meany',
       'A Thin Dark Line (Mysteries &amp; Horror)',
       "A Thousand Acres (Ballantine Reader's Circle)", 'A Time to Kill',
       "A Virtuous Woman (Oprah's Book Club (Paperback))",
       'A Walk to Remember', 'A Widow for One Year', 'A Wrinkle In Time',
  

In [346]:
book_list.to_csv('book_name.csv')

In [343]:
pickle.dump(book_list,open('book_list.pkl','wb'))

In [347]:
book_pivot.to_csv("book_pivot.csv")

In [764]:

def suggest1(book_name):
    l=[]
    book_index=np.where(book_pivot.index==book_name)[0][0]
    distances,suggestion=model.kneighbors(book_pivot.iloc[book_index,:].values.reshape(1, -1),n_neighbors=6)
    for i in range(len(suggestion)):
        l.append(book_pivot.index[suggestion[i]].values[1:6])
    return l

In [765]:
suggest1('1984')

[array(['No Safe Place', 'A Civil Action', "Foucault's Pendulum",
        'Long After Midnight', 'Abduction'], dtype=object)]

In [375]:
l

[array(['Exclusive', 'No Safe Place', 'Jacob Have I Loved',
        'Long After Midnight', 'The Cradle Will Fall'], dtype=object)]

In [377]:
for i in l[0]:
    print(i)

Exclusive
No Safe Place
Jacob Have I Loved
Long After Midnight
The Cradle Will Fall


In [378]:
l[0][0]

'Exclusive'

In [380]:
list(book_sparse)

[<1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 28 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 47 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 34 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 13 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 25 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 26 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 15 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 25 stored elements in Compressed Sparse Row format>,
 <1x888 sparse matrix of type '<class 'numpy.float64'>'
 	with 5 stored 

In [381]:
help(csr_matrix)

Help on class csr_matrix in module scipy.sparse.csr:

class csr_matrix(scipy.sparse.compressed._cs_matrix)
 |  csr_matrix(arg1, shape=None, dtype=None, copy=False)
 |  
 |  Compressed Sparse Row matrix
 |  
 |  This can be instantiated in several ways:
 |      csr_matrix(D)
 |          with a dense matrix or rank-2 ndarray D
 |  
 |      csr_matrix(S)
 |          with another sparse matrix S (equivalent to S.tocsr())
 |  
 |      csr_matrix((M, N), [dtype])
 |          to construct an empty matrix with shape (M, N)
 |          dtype is optional, defaulting to dtype='d'.
 |  
 |      csr_matrix((data, (row_ind, col_ind)), [shape=(M, N)])
 |          where ``data``, ``row_ind`` and ``col_ind`` satisfy the
 |          relationship ``a[row_ind[k], col_ind[k]] = data[k]``.
 |  
 |      csr_matrix((data, indices, indptr), [shape=(M, N)])
 |          is the standard CSR representation where the column indices for
 |          row i are stored in ``indices[indptr[i]:indptr[i+1]]`` and their
 | 

In [393]:
l=book_sparse.toarray().reshape(1,-1)
l

array([[9., 0., 0., ..., 0., 0., 0.]])

In [394]:
pickle.dump(l,open('book_sparse.pkl','wb'))

In [387]:
book_sparse.to_csv('book_sparse.csv')

AttributeError: to_csv not found

In [411]:
book_pivot

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [416]:
pickle.dump(book_pivot,open('book_pivot.pkl','wb'))

In [406]:
ss=open('book_pivot.pkl','rb')

In [408]:
ss=pd.read_pickle('book_pivot.pkl')

In [409]:
ss

title
0                                                 1984
1                                  1st to Die: A Novel
2                                           2nd Chance
3                                            4 Blondes
4                                84 Charing Cross Road
..                                                 ...
737                                    Year of Wonders
738                                   You Belong To Me
739  Zen and the Art of Motorcycle Maintenance: An ...
740                                               Zoya
741                                \O\" Is for Outlaw"

[742 rows x 1 columns]

In [417]:
ss=pickle.load(open('book_pivot.pkl','rb'))

In [418]:
ss

user_id                                             254     2276    2766    \
title                                                                        
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             2977    3363    3757    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             4017    4385    6242    \
title                                                                        
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
84 Charing Cross Road                                  0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     7.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

user_id                                             6251    ...  274004  \
title                                                       ...           
1984                                                   0.0  ...     0.0   
1st to Die: A Novel                                    0.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
84 Charing Cross Road                                  0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [419]:
pickle.dump(book_sparse,open('book_sparse.pkl','wb'))

In [420]:
ss=pickle.load(open('book_sparse.pkl','rb'))

In [424]:
ss

<742x888 sparse matrix of type '<class 'numpy.float64'>'
	with 14942 stored elements in Compressed Sparse Row format>

In [425]:
book_sparse

<742x888 sparse matrix of type '<class 'numpy.float64'>'
	with 14942 stored elements in Compressed Sparse Row format>

In [432]:
book_list.head(2)

title
0                 1984
1  1st to Die: A Novel

In [431]:
image.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [460]:
image[image['Book-Title']=='1984'][0:1]

ISBN Book-Title    Book-Author Year-Of-Publication    Publisher  \
2104  0451524934       1984  George Orwell                1990  Signet Book   

                                            Image-URL-S  \
2104  http://images.amazon.com/images/P/0451524934.0...   

                                            Image-URL-M  \
2104  http://images.amazon.com/images/P/0451524934.0...   

                                            Image-URL-L  
2104  http://images.amazon.com/images/P/0451524934.0...

In [478]:
image['Image-URL-M'][image['Book-Title']=='1984'][0:1]

2104    http://images.amazon.com/images/P/0451524934.0...
Name: Image-URL-M, dtype: object

In [472]:
image['Image-URL-M'][0]

'http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg'

In [461]:
image.shape

(271360, 8)

In [462]:
image=image[['ISBN','Book-Title','Image-URL-M']]
image.head(3)

ISBN            Book-Title  \
0  0195153448   Classical Mythology   
1  0002005018          Clara Callan   
2  0060973129  Decision in Normandy   

                                         Image-URL-M  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...

In [463]:
pickle.dump(image,open('book_image.pkl','wb'))

In [465]:
ll=pickle.load(open('book_image.pkl','rb'))
ll

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                                              Image-URL-M  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.amazon.com/images/P/0393045218.0...  
...                                                   ...  
271355  http://images.amazon.com/images/P/0440400988.0...  
271356  http://images.amazon.com/images/P/0525447644.0...  
271357  http://images.amazon.com/images/P/006008667X.0...  
271358  http://images.amazon.com/images/P/0192126040.0...  
271359  http://images.amazon.com/images/P/0767409752.0...  

[271360 rows x 3 columns]

In [601]:
ll['Image-URL-M'][0]

'http://images.amazon.com/images/P/0195153448.01.MZZZZZZZ.jpg'

In [482]:
image['ISBN'][image['Book-Title']=='1984'][0:1].values[0]

'0451524934'

In [483]:
image['ISBN'][image['Book-Title']=='1984'][0:1]

2104    0451524934
Name: ISBN, dtype: object

In [504]:
book_name=[i for i in book_list['title']]
ISBN=image[image['Book-Title'].isin([i for i in book_list['title']])]


In [511]:
ISBN.groupby('Book-Title')['Image-URL-M'].count().reset_index()

Book-Title  Image-URL-M
0                                                 1984            9
1                                  1st to Die: A Novel            2
2                                           2nd Chance            4
3                                            4 Blondes            1
4                                84 Charing Cross Road            3
..                                                 ...          ...
737                                    Year of Wonders            1
738                                   You Belong To Me            1
739  Zen and the Art of Motorcycle Maintenance: An ...            3
740                                               Zoya            2
741                                \O\" Is for Outlaw"            2

[742 rows x 2 columns]

In [488]:
[i for i in book_list['title']]

['1984',
 '1st to Die: A Novel',
 '2nd Chance',
 '4 Blondes',
 '84 Charing Cross Road',
 'A Bend in the Road',
 'A Case of Need',
 'A Child Called \\It\\": One Child\'s Courage to Survive"',
 'A Civil Action',
 'A Cry In The Night',
 'A Darkness More Than Night',
 'A Day Late and a Dollar Short',
 'A Fine Balance',
 'A Great Deliverance',
 'A Heartbreaking Work of Staggering Genius',
 'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
 'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
 'A Man Named Dave: A Story of Triumph and Forgiveness',
 'A Man in Full',
 'A Map of the World',
 'A Painted House',
 'A Patchwork Planet',
 'A Prayer for Owen Meany',
 'A Thin Dark Line (Mysteries &amp; Horror)',
 "A Thousand Acres (Ballantine Reader's Circle)",
 'A Time to Kill',
 "A Virtuous Woman (Oprah's Book Club (Paperback))",
 'A Walk to Remember',
 'A Widow for One Year',
 'A Wrinkle In Time',
 'A Wrinkle in Time',
 'A Year in Provence',
 "ANGELA'S ASHES",
 'Abduction',
 'Abou

In [514]:
image['ISBN'][image['Book-Title'] == 'Body of Lies'][0:1].values[0]

'0553582143'

In [543]:
def get_ISBN(book_name):
    ISBN=[]
    for i in book_name:
        ISBN.append(image['ISBN'][image['Book-Title'] == i][0:1].values[0])
    return ISBN

In [557]:
get_ISBN(['Bastard Out of Carolina','Deck the Halls (Holiday Classics)',
 'Degree of Guilt',
 'Deja Dead',
 'Desperation',
 'Devices and Desires',
 'Different Seasons',
 'Digital Fortress : A Thriller',
 'Disclosure'])

['0452269571',
 '0743418131',
 '034538184X',
 '0671011367',
 '0451188462',
 '0446359750',
 '0451124340',
 '0312995423',
 '0345391055']

In [517]:
get_ISBN('Bag of Bones')

'067102423X'

In [520]:
get_ISBN("Don't Sweat the Small Stuff and It's All Small Stuff : Simple Ways to Keep the Little Things from Taking Over Your Life (Don't Sweat the Small Stuff Series)")

'0786881852'

In [570]:
def get_ISBN(book_name):
    ISBN=[]
    for i in book_name:
        ISBN.append(image['ISBN'][image['Book-Title'] == i][0:1].values[0])
    return ISBN

def fetch_image(book_name):
    ISBN=[]
    for i in get_ISBN(book_name):
        ISBN.append(i)
        print("view-source:https://www.googleapis.com/books/v1/volumes?q=isbn:".format(ISBN))

In [571]:
fetch_image('Disclosure')

IndexError: index 0 is out of bounds for axis 0 with size 0

In [567]:
get_ISBN(['Bastard Out of Carolina','Deck the Halls (Holiday Classics)',
 'Degree of Guilt',
 'Deja Dead',
 'Desperation',
 'Devices and Desires',
 'Different Seasons',
 'Digital Fortress : A Thriller',
 'Disclosure'])

['0452269571',
 '0743418131',
 '034538184X',
 '0671011367',
 '0451188462',
 '0446359750',
 '0451124340',
 '0312995423',
 '0345391055']

In [ ]:
def get_ISBN(book_name):
    ISBN=[]
    for i in book_name:
        ISBN.append(image['ISBN'][image['Book-Title'] == i][0:1].values[0])
    return ISBN

In [656]:
image

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                                              Image-URL-M  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.amazon.com/images/P/0393045218.0...  
...                                                   ...  
271355  http://images.amazon.com/images/P/0440400988.0...  
271356  http://images.amazon.com/images/P/0525447644.0...  
271357  http://images.amazon.com/images/P/006008667X.0...  
271358  http://images.amazon.com/images/P/0192126040.0...  
271359  http://images.amazon.com/images/P/0767409752.0...  

[271360 rows x 3 columns]

In [663]:
ii=image.copy()

In [683]:
book_names_list=book_list['title'].values
book_names_list=list(book_names_list)
book_names_list

['1984',
 '1st to Die: A Novel',
 '2nd Chance',
 '4 Blondes',
 '84 Charing Cross Road',
 'A Bend in the Road',
 'A Case of Need',
 'A Child Called \\It\\": One Child\'s Courage to Survive"',
 'A Civil Action',
 'A Cry In The Night',
 'A Darkness More Than Night',
 'A Day Late and a Dollar Short',
 'A Fine Balance',
 'A Great Deliverance',
 'A Heartbreaking Work of Staggering Genius',
 'A Is for Alibi (Kinsey Millhone Mysteries (Paperback))',
 'A Lesson Before Dying (Vintage Contemporaries (Paperback))',
 'A Man Named Dave: A Story of Triumph and Forgiveness',
 'A Man in Full',
 'A Map of the World',
 'A Painted House',
 'A Patchwork Planet',
 'A Prayer for Owen Meany',
 'A Thin Dark Line (Mysteries &amp; Horror)',
 "A Thousand Acres (Ballantine Reader's Circle)",
 'A Time to Kill',
 "A Virtuous Woman (Oprah's Book Club (Paperback))",
 'A Walk to Remember',
 'A Widow for One Year',
 'A Wrinkle In Time',
 'A Wrinkle in Time',
 'A Year in Provence',
 "ANGELA'S ASHES",
 'Abduction',
 'Abou

In [685]:
ii=image[image['Book-Title'].isin(book_names_list)]
ii

ISBN                                      Book-Title  \
5       0399135782                          The Kitchen God's Wife   
18      0440234743                                   The Testament   
19      0452264464            Beloved (Plume Contemporary Fiction)   
26      0971880107                                     Wild Animus   
27      0345402871                                        Airframe   
...            ...                                             ...   
268974  0745168086                             The Handmaid's Tale   
269379  0582060184                                            1984   
269944  0805041494                                    Still Waters   
270659  0553290703                                       Lightning   
271169  0395647398  The Two Towers (The Lord of the Rings, Part 2)   

                                              Image-URL-M  
5       http://images.amazon.com/images/P/0399135782.0...  
18      http://images.amazon.com/images/P/0440234743.0...  
19      http://images.amazon.com/images/P/0452264464.0...  
26      http://images.amazon.com/images/P/0971880107.0...  
27      http://images.amazon.com/images/P/0345402871.0...  
...                                                   ...  
268974  http://images.amazon.com/images/P/0745168086.0...  
269379  http://images.amazon.com/images/P/0582060184.0...  
269944  http://images.amazon.com/images/P/0805041494.0...  
270659  http://images.amazon.com/images/P/0553290703.0...  
271169  http://images.amazon.com/images/P/0395647398.0...  

[2597 rows x 3 columns]

In [686]:
ii.nunique()

ISBN           2597
Book-Title      742
Image-URL-M    2555
dtype: int64

In [695]:
ii.reset_index(inplace=True)

In [696]:
ii

index        ISBN                                      Book-Title  \
0          5  0399135782                          The Kitchen God's Wife   
1         18  0440234743                                   The Testament   
2         19  0452264464            Beloved (Plume Contemporary Fiction)   
3         26  0971880107                                     Wild Animus   
4         27  0345402871                                        Airframe   
...      ...         ...                                             ...   
2592  268974  0745168086                             The Handmaid's Tale   
2593  269379  0582060184                                            1984   
2594  269944  0805041494                                    Still Waters   
2595  270659  0553290703                                       Lightning   
2596  271169  0395647398  The Two Towers (The Lord of the Rings, Part 2)   

                                            Image-URL-M  
0     http://images.amazon.com/images/P/0399135782.0...  
1     http://images.amazon.com/images/P/0440234743.0...  
2     http://images.amazon.com/images/P/0452264464.0...  
3     http://images.amazon.com/images/P/0971880107.0...  
4     http://images.amazon.com/images/P/0345402871.0...  
...                                                 ...  
2592  http://images.amazon.com/images/P/0745168086.0...  
2593  http://images.amazon.com/images/P/0582060184.0...  
2594  http://images.amazon.com/images/P/0805041494.0...  
2595  http://images.amazon.com/images/P/0553290703.0...  
2596  http://images.amazon.com/images/P/0395647398.0...  

[2597 rows x 4 columns]

In [687]:
def get_link_shorten(links):
    for i in range(len(links)):
        links[i]=links[i][0:-16]
    return links

In [688]:
link=ii['Image-URL-M']

In [698]:
get_link_shorten(ii['Image-URL-M'])

C:\Users\adila\AppData\Local\Temp/ipykernel_19632/4199766331.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  links[i]=links[i][0:-16]


0       http://images.amazon.com/images/P/0399135782
1       http://images.amazon.com/images/P/0440234743
2       http://images.amazon.com/images/P/0452264464
3       http://images.amazon.com/images/P/0971880107
4       http://images.amazon.com/images/P/0345402871
                            ...                     
2592    http://images.amazon.com/images/P/0745168086
2593    http://images.amazon.com/images/P/0582060184
2594    http://images.amazon.com/images/P/0805041494
2595    http://images.amazon.com/images/P/0553290703
2596    http://images.amazon.com/images/P/0395647398
Name: Image-URL-M, Length: 2597, dtype: object

In [699]:
ii

index        ISBN                                      Book-Title  \
0          5  0399135782                          The Kitchen God's Wife   
1         18  0440234743                                   The Testament   
2         19  0452264464            Beloved (Plume Contemporary Fiction)   
3         26  0971880107                                     Wild Animus   
4         27  0345402871                                        Airframe   
...      ...         ...                                             ...   
2592  268974  0745168086                             The Handmaid's Tale   
2593  269379  0582060184                                            1984   
2594  269944  0805041494                                    Still Waters   
2595  270659  0553290703                                       Lightning   
2596  271169  0395647398  The Two Towers (The Lord of the Rings, Part 2)   

                                       Image-URL-M  
0     http://images.amazon.com/images/P/0399135782  
1     http://images.amazon.com/images/P/0440234743  
2     http://images.amazon.com/images/P/0452264464  
3     http://images.amazon.com/images/P/0971880107  
4     http://images.amazon.com/images/P/0345402871  
...                                            ...  
2592  http://images.amazon.com/images/P/0745168086  
2593  http://images.amazon.com/images/P/0582060184  
2594  http://images.amazon.com/images/P/0805041494  
2595  http://images.amazon.com/images/P/0553290703  
2596  http://images.amazon.com/images/P/0395647398  

[2597 rows x 4 columns]

In [757]:
def get_ISBN(book_name):
    ISBN=[]
    for i in book_name:
        ISBN.append(ii['ISBN'][ii['Book-Title'] == i][0:1].values[0])
    return ISBN

def get_link(book_name):
    link=[]
    for i in get_ISBN(book_name):
        link.append(ii['Image-URL-M'][ii['ISBN'] == i][0:1].values[0])
    return link

        

In [758]:
ii['Image-URL-M'][ii['Book-Title'] == '2nd Chance'][0:1]

147    http://images.amazon.com/images/P/0316693200
Name: Image-URL-M, dtype: object

In [759]:
get_link(['1984',
 '1st to Die: A Novel',
 '2nd Chance'])


['http://images.amazon.com/images/P/0451524934',
 'http://images.amazon.com/images/P/0446610038',
 'http://images.amazon.com/images/P/0316693200']

In [707]:
get_ISBN(['1984',
 '1st to Die: A Novel',
 '2nd Chance'])

['0451524934', '0446610038', '0316693200']

In [760]:
get_link(book_names_list)

['http://images.amazon.com/images/P/0451524934',
 'http://images.amazon.com/images/P/0446610038',
 'http://images.amazon.com/images/P/0316693200',
 'http://images.amazon.com/images/P/0451203895',
 'http://images.amazon.com/images/P/0860074382',
 'http://images.amazon.com/images/P/0446527785',
 'http://images.amazon.com/images/P/0451183665',
 'http://images.amazon.com/images/P/1558743669',
 'http://images.amazon.com/images/P/0679772677',
 'http://images.amazon.com/images/P/0671886665',
 'http://images.amazon.com/images/P/0446667900',
 'http://images.amazon.com/images/P/0451204948',
 'http://images.amazon.com/images/P/140003065X',
 'http://images.amazon.com/images/P/0553278029',
 'http://images.amazon.com/images/P/0375725784',
 'http://images.amazon.com/images/P/0553279912',
 'http://images.amazon.com/images/P/0375702709',
 'http://images.amazon.com/images/P/0525945210',
 'http://images.amazon.com/images/P/0374270325',
 'http://images.amazon.com/images/P/0385500769',
 'http://images.amaz

In [ ]:
book_names_list

In [761]:
len(get_link(book_names_list))

742

In [762]:
pickle.dump(ii,open('book_image.pkl','wb'))

In [609]:
requests.get('http://images.amazon.com/images/P/0195153448')

<Response [200]>